apps/www/_blog/2024-12-06-database-build-v2.mdx

```diff
@@ -0,0 +1,218 @@
+
+If you’ve developed any browser app that connects to a backend API, you’ve likely experienced CORS. CORS stands for cross-origin-resource-sharing, and is a security mechanism built into browsers to prevent websites from connecting to APIs from a different domain. Quite often though there are legitimate reasons to connect to a different domain, and to support this, the server simply has to send back HTTP response headers that explicitly allow your app to connect to it.
+
+In the case of database.build, we need to connect directly to APIs like OpenAI’s https://api.openai.com/v1 or any other provider that you choose - and this will always be cross-origin since requests are coming from the browser. The good news is: most LLM providers add CORS headers by default which means we can connect directly to their APIs from the browser with no extra work. Without this, our only option would be to route every request through a backend proxy (which is not subject to CORS restrictions).
+
+### Service workers
+
+If you’ve ever dug into database.build's [source code](https://github.com/supabase-community/database-build), you’ll know that we heavily use Vercel’s [AI SDK](https://sdk.vercel.ai/docs/introduction) to simplify LLM streaming and client side tool calls. Vercel has built a great DX around inference by providing convenient hooks like `useChat()` to manage AI chat messages in React.
+
+The challenge with Vercel’s AI SDK for BYO-LLM is that the SDK expects a client-server API structure.` useChat()` connects to a server-side `/api/chat` route that is responsible for sending these messages downstream to the appropriate provider. In normal situations, this is the best architecture to protect API keys and custom logic on the server side. In our case though where users dynamically provide their own API keys, our preference is to send downstream requests directly from the browser. If we wish to continue using our existing `useChat()` infrastructure, we need a way to intercept requests to `/api/chat` and handle them ourselves in-browser.
+
+[Service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) to the rescue! One of the core features of service workers is the ability to intercept fetch requests and handle them client side - exactly what we need. With this approach, we can simply leave our `useChat()` logic as-is, then create a lightweight service worker that does the following:
+
+1. Detects whether you’re using your own model by reading local state from [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) (service workers don’t have access to local storage, so we opt for IndexedDB).
+1. If detected, it serves a local HTTP handler on `/api/chat`, sending the request directly to your provider’s API.
+1. Otherwise, it forwards the request to our backend as it did before.
+
+![database.build v2 BYO-LLM API key](/images/blog/launch-week-13/day-5-database-build-v2/llm-service-worker-dark.png)
```

Comment:
@fsansalvadore this too. need to make sure it looks good in both dark and light modes.